File: cs/AirtimeClient.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Net;
using Newtonsoft.Json;
using System.Security.Cryptography;
using System.Configuration;

namespace Cisl
{
    public class AirtimeClient
    {
        private String URL = String.Empty;
        private String ClientId = String.Empty;
        private String ClientKey = String.Empty;
        private String TerminalId = "AirtimeClient";

        public AirtimeClient()
        {
            this.URL = ConfigurationManager.AppSettings["URL"];
            this.ClientId = ConfigurationManager.AppSettings["ClientId"];
            this.ClientKey = ConfigurationManager.AppSettings["ClientKey"];
        }

        public AirtimeClient(String URL, String ClientId, String ClientKey)
        {
            this.URL = URL;
            this.ClientId = ClientId;
            this.ClientKey = ClientKey;
        }

        public AirtimeClient(String ClientId, String ClientKey)
        {
            this.URL = ConfigurationManager.AppSettings["URL"];
            this.ClientId = ClientId;
            this.ClientKey = ClientKey;
        }

        public static Network Parse(String net)
        {
            switch (net.ToUpper())
            {
                case "AIR":
                case "AIRTEL":
                    return Network.AIR;

                case "ETI":
                case "ETISALAT":
                    return Network.ETI;

                case "GLO":
                    return Network.GLO;

                case "MTN":
                    return Network.MTN;

                default:
                    throw new Exception("Unknown Network");
            }
        }

        public BalanceResult Balance()
        {
            String Nonce = DateTime.Now.ToString("yyyyMMddHHmmssfff");

            WebClient http = new WebClient();
            String Signature = ComputeSignature(Nonce, http.QueryString);

            http.Headers.Add("ClientId", ClientId);
            http.Headers.Add("Signature", Signature);
            http.Headers.Add("Nonce", Nonce);

            String response = http.DownloadString(URL + "/airtime/Balance");
            BalanceResult result = JsonConvert.DeserializeObject<BalanceResult>(response);
            return result;
        }

        public BalanceResult AgentBalance()
        {
            String Nonce = DateTime.Now.ToString("yyyyMMddHHmmssfff");

            WebClient http = new WebClient();
            String Signature = ComputeSignature(Nonce, http.QueryString);

            http.Headers.Add("AgentId", ClientId);
            http.Headers.Add("Signature", Signature);
            http.Headers.Add("Nonce", Nonce);
            http.Headers.Add("TerminalId", TerminalId);

            String response = http.DownloadString(URL + "/terminal/Balance");
            BalanceResult result = JsonConvert.DeserializeObject<BalanceResult>(response);
            return result;
        }

        public AirtimeResult Credit(Network net, String msisdn, int amount, string xref)
        {
            AirtimeResult result = new AirtimeResult();
            result.ResultCode = "0";
            String Nonce = DateTime.Now.ToString("yyyyMMddHHmmssfff");
            msisdn = Regex.Replace(msisdn, @"(?<PREFIX>^0)(?<TAIL>\d*)", @"234${TAIL}");

            try
            {
                WebClient http = new WebClient();
                http.QueryString.Add("net", net.ToString());
                http.QueryString.Add("msisdn", msisdn);
                http.QueryString.Add("amount", amount.ToString());
                http.QueryString.Add("xref", xref);

                String Signature = ComputeSignature(Nonce, http.QueryString);               

                http.Headers.Add("ClientId", ClientId);
                http.Headers.Add("Signature", Signature);
                http.Headers.Add("Nonce", Nonce);

                String response = http.DownloadString(URL + "/airtime/Credit");
                result = JsonConvert.DeserializeObject<AirtimeResult>(response);
                result.ResultCode = "200";
            }
            catch (WebException wex)
            {
                if (wex.Status == WebExceptionStatus.ProtocolError)
                {
                    HttpStatusCode httpStatus = ((HttpWebResponse)wex.Response).StatusCode;
                    result.ResultCode = httpStatus.ToString();
                    result.message = wex.Message;
                    result.xref = xref;
                }
            }

            return result;
        }

        public AirtimeResult AgentCredit(Network net, String msisdn, int amount, string xref)
        {
            AirtimeResult result = new AirtimeResult();
            result.ResultCode = "0";
            String Nonce = DateTime.Now.ToString("yyyyMMddHHmmssfff");
            msisdn = Regex.Replace(msisdn, @"(?<PREFIX>^0)(?<TAIL>\d*)", @"234${TAIL}");

            try
            {
                WebClient http = new WebClient();
                http.QueryString.Add("net", net.ToString());
                http.QueryString.Add("msisdn", msisdn);
                http.QueryString.Add("amount", amount.ToString());
                http.QueryString.Add("xref", xref);

                String Signature = ComputeSignature(Nonce, http.QueryString);                

                http.Headers.Add("AgentId", ClientId);
                http.Headers.Add("Signature", Signature);
                http.Headers.Add("Nonce", Nonce);
                http.Headers.Add("TerminalId", TerminalId);

                String response = http.DownloadString(URL + "/terminal/Credit");
                result = JsonConvert.DeserializeObject<AirtimeResult>(response);
                result.ResultCode = "200";
            }
            catch (WebException wex)
            {
                if (wex.Status == WebExceptionStatus.ProtocolError)
                {
                    result.ResultCode = ((HttpWebResponse)wex.Response).StatusCode.ToString();
                }
            }

            return result;
        }

        public CheckResult Check(String reference)
        {
            String Nonce = DateTime.Now.ToString("yyyyMMddHHmmssfff");
            String Xref = Guid.NewGuid().ToString("N");

            WebClient http = new WebClient();
            http.QueryString.Add("reference", reference);

            String Signature = ComputeSignature(Nonce, http.QueryString);

            http.Headers.Add("ClientId", ClientId);
            http.Headers.Add("Signature", Signature);
            http.Headers.Add("Nonce", Nonce);

            String response = http.DownloadString(URL + "/airtime/Check");
            CheckResult result = JsonConvert.DeserializeObject<CheckResult>(response);
            return result;
        }

        private string ComputeSignature(String Nonce, NameValueCollection query)
        {
            String signature = String.Empty;
            StringBuilder queryString = new StringBuilder();
            if (query.Count > 0)
            {
                queryString.Append("?");
                queryString.Append(query.GetKey(0));
                queryString.Append("=");
                queryString.Append(query[0]);

                for (int i = 1; i < query.Count; i++)
                {
                    queryString.Append("&");
                    queryString.Append(query.GetKey(i));
                    queryString.Append("=");
                    queryString.Append(query[i]);
                }
            }

            byte[] data = Encoding.UTF8.GetBytes(Nonce + queryString.ToString());
            byte[] secretKeyBytes = Convert.FromBase64String(ClientKey);

            using (HMACSHA256 hmac = new HMACSHA256(secretKeyBytes))
            {
                byte[] signatureBytes = hmac.ComputeHash(data);
                signature = Convert.ToBase64String(signatureBytes);
            }

            return signature;
        }
    }

    public enum Network
    {
        MTN, AIR, GLO, ETI
    }

    public class AirtimeResult
    {
        public String id { get; set; }
        public String xref { get; set; }
        public int balance { get; set; }
        public int charge { get; set; }
        public String message { get; set; }
        public String ResultCode { get; set; }
    }

    public class BalanceResult
    {
        public int balance { get; set; }
        public int discount { get; set; }
    }

    public class CheckResult
    {
        public String id { get; set; }
        public String msisdn { get; set; }
        public String net { get; set; }
        public int amount { get; set; }
        public String xref { get; set; }
        public DateTime stamp { get; set; }
    }
}

File: java/src/AirtimeClient.java
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.io.InputStream;
import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Date;
import java.util.UUID;
import java.text.SimpleDateFormat;
import java.text.MessageFormat;
import javax.json.Json;
import javax.json.JsonObject;
import javax.json.JsonReader;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.apache.commons.codec.binary.Base64;
import javax.xml.bind.DatatypeConverter;

public class AirtimeClient {

	private String url;
	private String clientId;
	private String clientKey;
	public static String TerminalId = "0001";

	public AirtimeClient(String url, String clientId, String clientKey) {
		this.url = url;
		this.clientId = clientId;
		this.clientKey = clientKey;
	}

	private String nonce() {
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddhhmmssSSS");
		Date now = new Date();
		return dateFormat.format(now);
	}

	private String sign(String nonce, String query) throws NoSuchAlgorithmException, InvalidKeyException {
		byte[] data = (nonce + query).getBytes();
		byte[] secretKeyBytes = Base64.decodeBase64(clientKey);
		Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
		SecretKeySpec secret_key = new SecretKeySpec(secretKeyBytes, "HmacSHA256");
		sha256_HMAC.init(secret_key);
		//return Base64.encodeBase64String(sha256_HMAC.doFinal(data));
		return DatatypeConverter.printBase64Binary(sha256_HMAC.doFinal(data));
	}

	public JsonObject balance() {

		JsonObject balance = null;

		try {
			URL rest = new URL(url + "/airtime/Balance");
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, "");

			connection.setRequestProperty("ClientId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}
			
			JsonReader jreader = Json.createReader(connection.getInputStream());
			balance = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return balance;
	}

	public JsonObject credit(String net, String msisdn, int amount, String xref) {

		JsonObject credit = null;

		try {
			String query = MessageFormat.format("?net={0}&msisdn={1}&amount={2}&xref={3}", net, msisdn, amount, xref);
			URL rest = new URL(url + "/airtime/Credit" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("ClientId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}

			JsonReader jreader = Json.createReader(connection.getInputStream());
			credit = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return credit;
	}

	public JsonObject dataCredit(String net, String msisdn, int amount, String xref) {

		JsonObject credit = null;

		try {
			String query = MessageFormat.format("?net={0}&msisdn={1}&amount={2}&xref={3}", net, msisdn, amount, xref);
			URL rest = new URL(url + "/data/Credit" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("ClientId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}

			JsonReader jreader = Json.createReader(connection.getInputStream());
			credit = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return credit;
	}

	public JsonObject check(String xref) {

		JsonObject check = null;

		try {
			String query = MessageFormat.format("?reference={0}", xref);
			URL rest = new URL(url + "/airtime/Check" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("ClientId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}
			
			JsonReader jreader = Json.createReader(connection.getInputStream());
			check = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return check;
	}

	public JsonObject agentBalance() {

		JsonObject balance = null;

		try {
			URL rest = new URL(url + "/terminal/Balance");
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, "");

			connection.setRequestProperty("AgentId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);
			connection.setRequestProperty("TerminalId", TerminalId);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}
			
			JsonReader jreader = Json.createReader(connection.getInputStream());
			balance = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return balance;
	}

	public JsonObject agentCredit(String net, String msisdn, int amount, String xref) {

		JsonObject credit = null;

		try {
			String query = MessageFormat.format("?net={0}&msisdn={1}&amount={2}&xref={3}", net, msisdn, amount, xref);
			URL rest = new URL(url + "/terminal/Credit" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("AgentId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);
			connection.setRequestProperty("TerminalId", TerminalId);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}

			JsonReader jreader = Json.createReader(connection.getInputStream());
			credit = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return credit;
	}

	public JsonObject agentDataCredit(String net, String msisdn, int amount, String xref) {

		JsonObject credit = null;

		try {
			String query = MessageFormat.format("?net={0}&msisdn={1}&amount={2}&xref={3}", net, msisdn, amount, xref);
			URL rest = new URL(url + "/terminal/data/Credit" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("AgentId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);
			connection.setRequestProperty("TerminalId", TerminalId);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}

			JsonReader jreader = Json.createReader(connection.getInputStream());
			credit = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return credit;
	}

	public JsonObject agentCheck(String xref) {

		JsonObject check = null;

		try {
			String query = MessageFormat.format("?reference={0}", xref);
			URL rest = new URL(url + "/terminal/Check" + query);
			HttpURLConnection connection = (HttpURLConnection) rest.openConnection();
			connection.setRequestMethod("GET");

			String nonce = nonce();
			String signature = sign(nonce, query);

			connection.setRequestProperty("AgentId", clientId);
			connection.setRequestProperty("Signature", signature);
			connection.setRequestProperty("Nonce", nonce);
			connection.setRequestProperty("TerminalId", TerminalId);

			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
				throw new RuntimeException("Failed : HTTP error code : " + connection.getResponseCode());
			}
			
			JsonReader jreader = Json.createReader(connection.getInputStream());
			check = jreader.readObject();
			connection.disconnect();

		} catch (MalformedURLException e) {
			e.printStackTrace();

		} catch (IOException e) {
			e.printStackTrace();

		} catch (InvalidKeyException e) {
			e.printStackTrace();
			
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			
		}

		return check;
	}

	public static void main(String[] args) {

		String url = "https://api.cowriesys.com:2443";
		String clientId = "ClientId";
		String clientKey = "ClientKey";

		String net = args[0]; // net can be AIR, ETI, GLO, MTN
		String msisdn = args[1]; // msisdn in international format 234 prefix
		int amount =  Integer.parseInt(args[2]); // minimum varies across networks
		String xref = UUID.randomUUID().toString(); // any unique reference

		AirtimeClient airtime = new AirtimeClient(url, clientId, clientKey);
		JsonObject balance = airtime.balance();
		System.out.println(balance.toString());
		
		/**
		JsonObject credit = airtime.credit(net, msisdn, amount, xref);
		System.out.println(credit.toString());
				
		JsonObject check = airtime.check(xref);
		System.out.println(check.toString());
		*/
	}
}
